### PR TITLE
Initial attempt to make local docker dev setup

### PR DESCRIPTION
**Pull request by einarf**
Not tested with on windows yet, but I hope it supports using `.` when mapping the host volume of the Misago project in the compose file. Otherwise we need to use an absolute path.

`docker-compose up -d` and Misago should be available at http://localhost:8000`

Should probably also create a superuser, but wanted to share this before I added more stuff.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -47,6 +47,9 @@ pylint.txt
 .project
 .pydevproject
 
+# pycharm
+.idea/
```
Comment (rafalp) [inline .gitignore]: Please rename this section to "IDEs" and move `.vscode` there.

File: extras/wait_for_postgres.sh [hunk 2]
```diff
@@ -0,0 +1,8 @@
+#!/usr/bin/env bash
+export PGPASSWORD=$POSTGRES_PASSWORD
+RETRIES=10
+
+until psql -h $POSTGRES_HOST -U $POSTGRES_USER -d $POSTGRES_DB -c "select 1" > /dev/null 2>&1 || [ $RETRIES -eq 0 ]; do
+  echo "Waiting for postgres server, $((RETRIES--)) remaining attempts..."
+  sleep 5
+done
```
Comment (rafalp) [inline extras/wait_for_postgres.sh]: Is this file necessary? I've worked with quite a few docker setups and never encountered anything like that.

File: extras/createdevproject.py [hunk 3]
```diff
@@ -0,0 +1,47 @@
+"""
+Modifies the project template to fit with local development
+"""
+import os
```
Comment (rafalp) [inline extras/createdevproject.py]: There should be a blank line separating module-level docstring from first import.

File: extras/createdevproject.py [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+import sys
+from misago.core import setup
```
Comment (rafalp) [inline extras/createdevproject.py]: There should be a blank line separating package imports from stdlib imports.

File: extras/createdevproject.py [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+
+
+PROJECT_NAME = 'forum'
+
+BASE_DIR = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
+
+
+def main():
+    global PROJECT_NAME, SETTINGS_FILE
```
Comment (rafalp) [inline extras/createdevproject.py]: Why is global used here? Especially `PROJECT_NAME`  should be in scope.

File: extras/createdevproject.py [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+
+    # Avoid recreating if already present
+    if os.path.exists(settings_file):
+        return
+
+    setup.start_misago_project()
+    tweak_settings(settings_file)
+
+
+def tweak_settings(f):
```
Comment (rafalp) [inline extras/createdevproject.py]: Lets rename this to `fill_in_settings`. `tweak_settings` doesn't seem to be truthful to function's contents.

File: .gitignore [hunk 7]
```diff
@@ -63,9 +66,11 @@ pylint.txt
 _book/
 
 # Local development stuff
+forum/
 dev/
 dev-manage.py
 .vscode/
+env/
```
Comment (rafalp) [inline .gitignore]: What is `env` directory used for?

File: docker-compose.yaml [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+    image: postgres:9
+    environment:
+      - POSTGRES_USER=misago
+      - POSTGRES_PASSWORD=misago
+      - POSTGRES_DB=misago
+    volumes:
+      # To keep things simple we map the postgres data into a docker volume.
+      # When re-creating the database you need to delete the volume with `docker volume rm`.
+      # You can also map the path to the host if this is more convenient
+      - postgres:/var/lib/postgresql/data
```
Comment (rafalp) [inline docker-compose.yaml]: There's no need for postgresql volume in dev IMHO, we already have tools that let dev to quickly fill the database with mock data if they wish so.

File: extras/createdevproject.py [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+def main():
+    global PROJECT_NAME, SETTINGS_FILE
+
+    # Allow for overriding project name
+    if len(sys.argv) > 1:
+        PROJECT_NAME = sys.argv[1]
+    else:
+        sys.argv.append(PROJECT_NAME)
+
+    settings_file = os.path.join(BASE_DIR, PROJECT_NAME, PROJECT_NAME, 'settings.py')
```
Comment (rafalp) [inline extras/createdevproject.py]: Lets not create deep nesting for those files. Its fine to put `avatargallery`, `media`, `manage.py` etc. ect. in main dir, just make sure those are in `.gitignore` too.

File: .gitignore (hunk 7 above)
Comment (einarf) [reply .gitignore]: It's just another alternative for `venv` that apparently some plugin on pycharm makes.

Comment (rafalp) [reply .gitignore]: Ok

File: docker-compose.yaml (hunk 8 above)
Comment (einarf) [reply docker-compose.yaml]: That's a good point. We can just let the data die with the container.

File: extras/createdevproject.py (hunk 5 above)
Comment (einarf) [reply extras/createdevproject.py]: Will remove. It's remains of some early sins when testing!

File: extras/createdevproject.py (hunk 9 above)
Comment (einarf) [reply extras/createdevproject.py]: Not entirely sure if I understand. I'm just reading a settings file created by misago's `startproject` version.

Comment (rafalp) [reply extras/createdevproject.py]: I'm saying that there should be additional steps involved:

`rm forum/cron.txt`
`rm -rf forum/static`
`mv forum/manage.py manage.py`
`mv forum/avatargallery avatargallery`
`mv forum/media media`
`rm -rf forum/theme`

**Comment by rafalp** [comment]
Also, we should clear `STATICFILES_DIRS` so its `STATICFILES_DIRS = []`, and change email backend to console:

https://docs.djangoproject.com/en/2.0/topics/email/#console-backend

File: extras/wait_for_postgres.sh (hunk 2 above)
Comment (einarf) [reply extras/wait_for_postgres.sh]: It's a very common problem that postgres is not ready before django starts attempting to connect.

Comment (rafalp) [reply extras/wait_for_postgres.sh]: This should be a comment in that file then.

File: .gitignore [hunk 10]
```diff
@@ -75,3 +79,11 @@ venv36/
 future-manage.py
 misago-admin.py
 db.sqlite3
+
+# Local development files
+/manage.py
+/forum/
+/avatargallery/
+/static/
+/theme/
```
Comment (rafalp) [inline .gitignore]: `theme` and `cron` files should be removed from gitignore if they are also removed on deployment.

File: Dockerfile [hunk 11]
```diff
@@ -0,0 +1,28 @@
...
+    locales \
+    cron \
+    postgresql-client
+
+# Add requirements and install them. We do this unnecessasy rebuilding.
+ADD requirements.txt /
+RUN pip install -r requirements.txt
+
+# Add project source and run
+# ADD . /srv/misago (We don't need this in development, leaving out for now)
```
Comment (rafalp) [inline Dockerfile]: If its not needed, then lets not leave it commented out to make people wonderif maybe its needed now ;)

File: extras/createdevproject.py [hunk 12]
```diff
@@ -0,0 +1,51 @@
...
+        s = fd.read()
+
+        # Postgres
+        s = s.replace("'NAME': '',", "'NAME': os.environ['POSTGRES_DB'],")
+        s = s.replace("'USER': '',", "'USER': os.environ['POSTGRES_USER'],")
+        s = s.replace("'PASSWORD': '',", "'PASSWORD': os.environ['POSTGRES_PASSWORD'],")
+        s = s.replace("'HOST': 'localhost',", "'HOST': os.environ['POSTGRES_HOST'],")
+
+        # Specify console backend for email
+        s += "\nEMAIL_BACKEND = 'django.core.mail.backends.console.EmailBackend'\n"
```
Comment (rafalp) [inline extras/createdevproject.py]: We are deleting `theme` dir that is used in `STATICFILES_DIRS` setting, so that one needs to be emptied here too.
